editor/src/components/canvas/grid-panels-state.tsx

```diff
@@ -0,0 +1,223 @@
+ */
+export function wrapAroundColIndex(index: number): number {
+  const normalized = normalizeColIndex(index)
+  if (normalized >= IndexOfCanvas) {
+    return normalized - NumberOfColumns
+  } else {
+    return normalized
+  }
+}
+
+/**
+ * Normalizes the index to 0,1,2,3
+ */
+function normalizeColIndex(index: number): number {
+  return mod(index, NumberOfColumns)
+}
+
+export function updateLayout(
+  stored: StoredLayout,
+  paneToMove: StoredPanel, // must be referentially equal to the stored panel!
+  update: LayoutUpdate,
+): StoredLayout {
+  const panelToInsert = storedPanel(paneToMove)
+
+  function insertPanel(layout: StoredLayout) {
+    if (update.type === 'before-column' || update.type === 'after-column') {
+      const atLeastOneEmptyColumn = layout.some((col) => col.length === 0)
+      if (!atLeastOneEmptyColumn) {
+        // the user wants to create a new column and fill it with the moved Panel.
+        // if there's zero empty columns, it means we cannot create a new column, so we must bail out
+
+        return layout // BAIL OUT! TODO we should show a Toast
+      }
+      const newColumn: Array<StoredPanel> = [panelToInsert]
+
+      const normalizedIndex = normalizeColIndex(update.columnIndex)
+
+      const indexInArray = update.type === 'before-column' ? normalizedIndex : normalizedIndex + 1
+
+      const rightHandSide = normalizedIndex >= IndexOfCanvas
+
+      const withElementInserted = insert(indexInArray, newColumn, layout)
+      const withOldPanelRemoved = removeOldPanel(withElementInserted)
+
+      const indexOfFirstEmptyColumn = rightHandSide
+        ? findLastIndex(withOldPanelRemoved, (col) => col.length === 0)
+        : withOldPanelRemoved.findIndex((col) => col.length === 0)
+
+      return removeIndexFromArray(indexOfFirstEmptyColumn, withOldPanelRemoved)
+    }
+    if (update.type === 'before-index') {
+      const working = [...layout]
+
+      // insert
+      working[update.columnIndex] = insert(
+        update.indexInColumn,
+        panelToInsert,
+        working[update.columnIndex],
+      )
+
+      return removeOldPanel(working)
+    }
+    if (update.type === 'after-index') {
+      const working = [...layout]
+
+      // insert
+      working[update.columnIndex] = insert(
+        update.indexInColumn + 1,
+        panelToInsert,
+        working[update.columnIndex],
+      )
+
+      return removeOldPanel(working)
+    }
+
+    assertNever(update)
+  }
+
+  function removeOldPanel(layout: StoredLayout) {
+    return layout.map((column) => {
+      return removeAll(column, [paneToMove], (l, r) => l.uid === r.uid)
+    })
+  }
+
+  function floatColumnsTowardsEdges(layout: StoredLayout) {
+    const leftSide = layout.slice(0, IndexOfCanvas)
+    const rightSideReversed = layout.slice(IndexOfCanvas).reverse()
+    const leftSideFixed = accumulate(new Array(leftSide.length).fill([]), (acc) => {
+      let iterator = 0
```

Comment:
It's definitely not an iterator!
```suggestion
      let index = 0
```

Reply:
it's also not an index of the loop - it's the index of the "cursor" of the array the loop is optionally updating

Reply:
I renamed it to "indexInAccumulator"